### ISSUE TITLE
Inefficient SQL generated for Bulk Update from in-memory collection

### ISSUE THREAD

**Issue by Nate129** [linq2db/linq2db#2844]
Hi,

I am attempting to design a generic 'bulk update' method which uses a temp table to store the externally updated rows and joins the temp table with the actual table to do the update operation efficiently: 

```c#
public void UpdateRange<T>(IEnumerable<T> source) where T : class, IEntity
{
    using var connection = _context.CreateLinqToDbConnection();

    connection.BeginTransaction();

    using var tempTable = connection.CreateTempTable(source, tableName: $"{typeof(T).Name}UpdateTemp");

    var entityTable = connection.GetTable<T>();

    var updates = from e in entityTable
        join te in tempTable on e.Id equals te.Id into gj
        from te in gj
        select te;

    updates.Update(entityTable, e => e);

    connection.CommitTransaction();
}
```

I use this strategy as opposed to a `MERGE` statement to support RDBMS that don't support `MERGE`.

When working with the following test model and interfaces:

```c#
public interface IEntity
{
    object Id { get; set; }
}

public interface IEntity<T> : IEntity
{
    new T Id { get; set; }
}

public abstract class Entity<T> : IEntity<T>
{
    public T Id { get; set; }

    object IEntity.Id
    {
        get => Id;
        set => Id = value != null ? (T) value : default;
    }
}

public class Person : Entity<int>
{
    public string FirstName { get; set; }
    public string LastName { get; set; }
}
```
The following SQL is generated:

```SQL
UPDATE
	[People]
SET
	[Id] = (
		SELECT
			[gj].[Id]
		FROM
			[People] [e]
				INNER JOIN [PersonUpdateTemp] [gj] ON [e].[Id] = [gj].[Id]
		WHERE
			[People].[Id] = [e].[Id]
	),
	[LastName] = (
		SELECT
			[gj_1].[LastName]
		FROM
			[People] [e_1]
				INNER JOIN [PersonUpdateTemp] [gj_1] ON [e_1].[Id] = [gj_1].[Id]
		WHERE
			[People].[Id] = [e_1].[Id]
	),
	[FirstName] = (
		SELECT
			[gj_2].[FirstName]
		FROM
			[People] [e_2]
				INNER JOIN [PersonUpdateTemp] [gj_2] ON [e_2].[Id] = [gj_2].[Id]
		WHERE
			[People].[Id] = [e_2].[Id]
	)
WHERE
	EXISTS(
		SELECT
			*
		FROM
			[People] [e_3]
				INNER JOIN [PersonUpdateTemp] [gj_3] ON [e_3].[Id] = [gj_3].[Id]
		WHERE
			[People].[Id] = [e_3].[Id]
	)

```

I was hoping to have a single join and update statement updating each column from the join result. How would I accomplish this? 

I have a repo that contains a test that produces the above SQL [here](https://github.com/Nate129/Linq2DbTests/tree/master) (see the `Linq2DbUpdateTests` class). I retrieved the SQL by setting a breakpoint in the `TestService` class on the last line of the method `connection.CommitTransaction();` and inspecting the Command property's SQLText property.

### Environment details
linq2db version: 3.2.3
Database Server: Sqlite
Database Provider: EF Core (via Linq2db.EntityFrameworkCore
Operating system: Windows 10
.NET Framework: .NET Core

**Comment by sdanyliv** [comment]
Actually I don't know better way how to update records for SQLite. Do you have information about performant SQLite syntax in such case ?

**Comment by Nate129** [comment]
A join per column seems unacceptably slow for larger collections. I don't believe this is the fault of linq2db- I think this is a limitation of Sqlite. 

I suppose I'd need a different approach that would be efficient for Sqlite and other RBDMS as well.

I found this [stack overflow post](https://stackoverflow.com/questions/3845718/update-table-values-from-another-table-with-the-same-user-name/63079219#63079219) that seems to have a potential solution for merging two tables in Sqlite which looks kind of like what I was hoping to achieve.

The result might look something like (shorthand): 

```SQL
UPDATE [People]
SET [Id] = [PersonUpdateTemp].[Id],
       [LastName] = [PersonUpdateTemp].[LastName],
       [FirstName] = [PersonUpdateTemp].[FirstName],
FROM  [PersonUpdateTemp]
WHERE  [PersonUpdateTemp].[Id] = [People].[Id];
```

The only caveat is that `UPDATE FROM` is only available in Sqlite >= 3.3.3. [Here](https://sqlite.org/lang_update.html#upfrom) are the relevant Sqlite docs. Since EF Core supports Sqlite >= 3.7, that shouldn't be an issue, provided Linq2db can support it.

I'm having some trouble getting Linq2db to generate this. 

I tried:

```c#
var updates = from e in entityTable from te in tempTable where te.Id == e.Id select te;

updates.Update(entityTable, e => e);
```

which generates: 

```SQL
SELECT
	[te].[Id],
	[te].[LastName],
	[te].[FirstName]
FROM
	[PersonUpdateTemp] [te],
	[People] [e]
WHERE
	[te].[Id] = [e].[Id]
```

This SQL looks good for the first line. However, when running the second line, it throws the exception `Operation is not valid due to the current state of the object`.

Does Linq2db support `UPDATE FROM` in Sqlite?

**Comment by sdanyliv** [comment]
I have written this UPDATE for SQLite approx 4 years ago. So it’s time to review this part.

**Comment by Nate129** [comment]
Any updates on your progress reviewing `UPDATE` with Sqlite? I've been working on the update section of my queries again and have been running into this issue and others trying to get bulk updates from a memory collection work efficiently with Sqlite.

Thanks!

**Comment by MaceWindu** [comment]
This will be partially fixed by #3339 with `SET row_value = query`. 
UPDATE FROM not yet implemented (supported since sqlite 3.33.0 )

**Comment by MaceWindu** [comment]
Actually, we will try to fix it for next release too
